service/history/queue/timer_queue_processor.go

```diff
@@ -338,19 +370,23 @@ func (t *timerQueueProcessor) completeTimerLoop() {
 					break
 				}
 
-				t.logger.Error("Error complete timer task", tag.Error(err))
+				t.logger.Error("Failed to complete timer task", tag.Error(err))
 				if err == shard.ErrShardClosed {
-					go t.Stop()
+					if !t.shard.GetConfig().QueueProcessorEnableGracefulSyncShutdown() {
+						go t.Stop()
+						return
+					}
+
+					t.Stop()
 					return
 				}
-				backoff := time.Duration(attempt * 100)
-				time.Sleep(backoff * time.Millisecond)
 
 				select {
 				case <-t.shutdownChan:
-					// break the retry loop if shutdown chan is closed
-					break
-				default:
+					t.drain()
```

Comment:
do you think this might need a flipr guard too?

Reply:
I don't think this needs dynamicconfig guard. previous version also attempts to make that drain operation after breaking here. I'm just making it explicit and returning here so the execution doesn't reevaluate the top level select. also getting rid of time.sleep in favor of time.after so it would be more responsive to shutdown signal